pkg/roachprod/blobfixture/registry_test.go

```diff
@@ -0,0 +1,185 @@
+			createdAt:      makeTime(-1),
+			survivesGC:     true,
+			isLatestOfKind: false,
+		},
+		{
+			// This fixture is older than 2 days, but it is the latest of its kind,
+			// so it will not be garbage collected.
+			kind:           "kind-singleton",
+			createdAt:      makeTime(-10),
+			readyAt:        makeTime(-5),
+			survivesGC:     true,
+			isLatestOfKind: true,
+		},
+		{
+			// This fixture was obsolete for more than two days, so it will be
+			// deleted.
+			kind:           "kind-multiple",
+			createdAt:      makeTime(-10),
+			readyAt:        makeTime(-9),
+			survivesGC:     false,
+			isLatestOfKind: false,
+		},
+		{
+			// This fixture was not obsolete for more than one day, so it will not be
+			// deleted.
+			kind:           "kind-multiple",
+			createdAt:      makeTime(-5),
+			readyAt:        makeTime(-4),
+			survivesGC:     true,
+			isLatestOfKind: false,
+		},
+		{
+			// This fixture was recently created, so its predecessor will not be deleted.
+			kind:           "kind-multiple",
+			createdAt:      makeTime(-2),
+			readyAt:        makeTime(-0.5),
+			survivesGC:     true,
+			isLatestOfKind: true,
+		},
+		{
+			// This is the most recent fixture of its kind, but its not ready yet, so
+			// its not the latest and its not eligible for GC.
+			kind:           "kind-multiple",
+			createdAt:      makeTime(-1),
+			survivesGC:     true,
+			isLatestOfKind: false,
+		},
+	}
+
+	type fixturesCreated struct {
+		fixture  fixture
+		metadata FixtureMetadata
+	}
+
+	var created []fixturesCreated
+
+	var now time.Time
+	registry := Registry{
+		clock: func() time.Time {
+			return now
+		},
+		storage: nodelocal.TestingMakeNodelocalStorage(t.TempDir(), cluster.MakeTestingClusterSettings(), cloudpb.ExternalStorage{}),
+		uri: url.URL{
+			Scheme: "nodelocal",
+			Host:   "1",
+			Path:   "roachtest/v25.1",
+		},
+	}
+
+	lcfg := logger.Config{
+		Stdout: io.Discard,
+		Stderr: io.Discard,
+	}
+	l, err := lcfg.NewLogger("")
+	require.NoError(t, err)
+
+	ctx := context.Background()
+	for _, f := range fixtures {
+		now = f.createdAt
+
+		handle, err := registry.Create(ctx, f.kind, l)
+		require.NoError(t, err)
+
+		metadata := handle.Metadata()
+
+		writer, err := registry.storage.Writer(ctx, path.Join(metadata.DataPath, "sentinel"))
+		require.NoError(t, err)
+
+		_, err = writer.Write([]byte(metadata.CreatedAt.String()))
+		require.NoError(t, err)
+
+		require.NoError(t, writer.Close())
+
+		if !f.readyAt.IsZero() {
+			now = f.readyAt
+			require.NoError(t, handle.SetReadyAt(ctx))
+		}
+
+		created = append(created, fixturesCreated{
+			fixture:  f,
+			metadata: metadata,
+		})
+	}
+
+	now = makeTime(0)
+
+	require.NoError(t, registry.GC(ctx, l))
+
+	for _, c := range created {
+		metadata, err := registry.GetLatest(ctx, c.fixture.kind)
+		if c.fixture.isLatestOfKind {
+			require.NoError(t, err)
+			require.Equal(t, c.metadata.DataPath, metadata.DataPath)
+		} else {
+			if err != nil {
```

Comment:
we would only expect an error here if the fixture did not survive gc, right? I think we could assert that, similar to what you do for the data path below.

Reply:
No. This error indicates that the kind has no ready fixtures. The leaked fixtures may have been deleted by the gc or they may have been recent enough to be spared.

I agree this assert is a little weird. The test may be a little clearer if the inputs were grouped by kind, then I could define an assertion that applies to the kind instead of an individual fixture instance.